Il2CppInspector.Common/Outputs/Namespace.cs

```diff
@@ -0,0 +1,67 @@
+﻿/*
+    Copyright 2017-2020 Katy Coe - http://www.hearthcode.org - http://www.djkaty.com
+    Copyright 2020 Robert Xiao - https://robertxiao.ca
+
+    All rights reserved.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Il2CppInspector.Outputs
+{
+    /// <summary>
+    /// A utility class for managing names in a common namespace.
+    /// </summary>
+    public class Namespace
+    {
+        private readonly Dictionary<string, int> renameCount = new Dictionary<string, int>();
+
+        public void ReserveName(string name) {
+            if (renameCount.ContainsKey(name)) {
+                throw new Exception($"Can't reserve {name}: already taken!");
+            }
+            renameCount[name] = 0;
+        }
+
+        public Namer<T> MakeNamer<T>(Namer<T>.KeyFunc keyFunc) {
+            return new Namer<T>(this, keyFunc);
+        }
+
+        /// <summary>
+        /// A class for managing objects of a common type within a namespace.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        public class Namer<T>
+        {
+            private Namespace ns;
+            private readonly Dictionary<T, string> names = new Dictionary<T, string>();
+            public delegate string KeyFunc(T t);
+            private readonly KeyFunc keyFunc;
+
+            public Namer(Namespace ns, KeyFunc keyFunc) {
+                this.ns = ns;
+                this.keyFunc = keyFunc;
+            }
+
+            public string GetName(T t) {
+                string name;
+                if (names.TryGetValue(t, out name))
+                    return name;
+                name = keyFunc(t);
+                // This approach avoids linear scan (quadratic blowup) if there are a lot of similarly-named objects.
+                if (ns.renameCount.ContainsKey(name)) {
+                    int v = ns.renameCount[name] + 1;
+                    while (ns.renameCount.ContainsKey(name + "_" + v))
+                        v++;
+                    ns.renameCount[name] = v;
+                    name = name + "_" + v;
+                }
+                ns.renameCount[name] = 0;
+                names[t] = name;
+                return name;
+            }
```

Comment:
At some point I was mulling making CppDeclarations an actual output module too, capable of outputting a C/C++ header file. Might be worth doing that in the near future?

Reply:
I guess you're thinking about #37 so yes, we can do that, however since the IDA output uses CppDeclarations I think it's better to move it to a new folder and create a new output module like "CppHeaders" that uses it in the same way the IDA output does. Also in respect of the discussion of offset resolution below I'm guessing we will make something like a new "CppType" that has a collection of "CppField", so having them all in their own Cpp folder just seems neater to me.